### PR TITLE
azure data factory resource implementation : singular & plural resources

### PR DESCRIPTION
**Pull request by sa-progress**
### Description

Please describe what this change achieves. Ensure you have read the [Contributing to InSpec](https://github.com/inspec/inspec-azure/CONTRIBUTING.MD) document before submitting.

### Issues Resolved

List any existing issues this PR resolves, or any Discourse or StackOverflow discussion that's relevant

### Check List

- [x] New functionality includes tests
- [x] All tests pass
- [x] `rake lint` passes
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

**Comment by soumyo13** [comment]
DCO is failing because while committing you have to signoff. Use the command git commit -s -m "branch_name".

File: docs/resources/azure_data_factories.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+describe azure_data_factories(resource_group: 'my-rg') do
+  #...
+end
+```
+
+## Properties
+
+|Property       | Description                                                                          | Filter Criteria<superscript>*</superscript> |
+|---------------|--------------------------------------------------------------------------------------|-----------------|
+| name           | A list of the unique resource ids.                                                   | `name`            |
```
Comment (soumyo13) [inline docs/resources/azure_data_factories.md:46]: As this is a plural resource, we have to change the field name from name to names. And the plural resource, in the filtertable you have written names

Comment (soumyo13) [reply docs/resources/azure_data_factories.md:46]: Please also add the following fields that you have mentioned in the filter table types, provisioning_states and locations.

File: docs/resources/azure_data_factories.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+```
+or
+```ruby
+describe azure_data_factories(resource_group: 'my-rg') do
+  #...
+end
+```
+
+## Properties
+
```
Comment (soumyo13) [inline docs/resources/azure_data_factories.md:43]: Please add the required parameters.

File: docs/resources/azure_data_factory.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  its('type') { should eq 'Microsoft.DataFactory/factories' }
+end
+```
+
+## Parameters
+
+| Name                           | Description                                                                      |
+|--------------------------------|----------------------------------------------------------------------------------|
+| resource_group                 | Azure resource group that the targeted resource resides in. `MyResourceGroup`    |
+| name                           | Name of the Azure resource to test. `MyDf`                                       |
```
Comment (soumyo13) [inline docs/resources/azure_data_factory.md:45]: Please add the other fields.

File: libraries/azure_data_factories.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  name 'azure_data_factories'
+  desc 'List azure data factories'
+  example <<-EXAMPLE
+    describe azure_data_factories(resource_group: 'example') do
+      it { should exist }
+    end
+  EXAMPLE
+
+  def initialize(opts = {})
+    # Options should be Hash type. Otherwise Ruby will raise an error when we try to access the keys.
```
Comment (soumyo13) [inline libraries/azure_data_factories.rb:13]: Can we remove the comments if it is not required from the plural resource?

File: libraries/azure_data_factory.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  example <<-EXAMPLE
+    describe azure_data_factory(resource_group: 'example', name: 'factoryName') do
+      it { should have_monitoring_agent_installed }
+    end
+  EXAMPLE
+
+  def initialize(opts = {})
+    # Options should be Hash type. Otherwise Ruby will raise an error when we try to access the keys.
+    raise ArgumentError, 'Parameters must be provided in an Hash object.' unless opts.is_a?(Hash)
+    # Azure REST API endpoint URL format for the resource:
```
Comment (soumyo13) [inline libraries/azure_data_factory.rb:15]: Can we remove the comments if it is not required from the singular resource?

File: test/integration/verify/controls/azure_data_factory.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+resource_group = input('resource_group', value: nil)
+factory_name = input('df_name', value: nil)
+df_location = input('df_location', value: nil)
+
+control 'azure_data_factory' do
+
```
Comment (soumyo13) [inline test/integration/verify/controls/azure_data_factory.rb:6]: Please add the description and the impact statements in the control block.

File: terraform/outputs.tf [hunk 7]
```diff
@@ -371,3 +371,11 @@ output "policy_definition_associated_cosmodb_id" {
 //  description = "the name of the policy exemption"
 //  value = azurerm_policy_exemption.inspec_compliance_policy_exemption.name
```
Comment (soumyo13) [inline terraform/outputs.tf:372]: If the commented lines are not required. Can we remove it?

**Review by soumyo13 — DISMISSED** [review]
Please check. I have left out some comments.

File: libraries/azure_data_factories.rb [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    # This will ensure constructing resource and passing `should_not exist` test.
+    return [] if @resources.empty?
+
+    @resources.each do |resource|
+      @table << {
+        id: resource[:id],
+        name: resource[:name],
+        type: resource[:type],
+        tags: resource[:tags],
+        provisioningState: resource[:properties][:provisioningState],
```
Comment (sathish-progress) [inline libraries/azure_data_factories.rb:90]: why not other properties?

File: docs/resources/azure_data_factory.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+## Azure REST API version, endpoint and http client parameters
+
+This resource interacts with api versions supported by the resource provider.
+The `api_version` can be defined as a resource parameter.
+If not provided, the latest version will be used.
+For more information, refer to [`azure_generic_resource`](azure_generic_resource.md).
+
+Unless defined, `azure_cloud` global endpoint, and default values for the http client will be used.
+For more information, refer to the resource pack [README](../../README.md).
+For api related info : [`Azure Data Factories Docs`](https://docs.microsoft.com/en-us/azure/data-factory/quickstart-create-data-factory-rest-api#create-a-data-factory).
```
Comment (sathish-progress) [inline docs/resources/azure_data_factory.md:19]: I guess this URL is not the right one

File: docs/resources/azure_data_factories.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+## Azure REST API version, endpoint and http client parameters
+
+This resource interacts with api versions supported by the resource provider.
+The `api_version` can be defined as a resource parameter.
+If not provided, the latest version will be used.
+For more information, refer to [`azure_generic_resource`](azure_generic_resource.md).
+
+Unless defined, `azure_cloud` global endpoint, and default values for the http client will be used.
+For more information, refer to the resource pack [README](../../README.md). 
+For api related info : [`Azure Data Factories Docs`](https://docs.microsoft.com/en-us/azure/data-factory/quickstart-create-data-factory-rest-api#create-a-data-factory).
```
Comment (sathish-progress) [inline docs/resources/azure_data_factories.md:19]: I guess this URL is not the right one

Comment (sa-progress) [reply docs/resources/azure_data_factories.md:19]: Yes Sateesh.I have updated the link.

File: libraries/azure_data_factories.rb (hunk 8 above)
Comment (sa-progress) [reply libraries/azure_data_factories.rb:90]: Hi Sateesh 
I will add etag property apart from that others are optional so it doesn't have any impact.
Providing json response for reference . 

{
  "name": "exampleFactoryName",
  "id": "/subscriptions/12345678-1234-1234-1234-12345678abc/resourceGroups/exampleResourceGroup/providers/Microsoft.DataFactory/factories/exampleFactoryName",
  "type": "Microsoft.DataFactory/factories",
  "properties": {
    "provisioningState": "Succeeded",
    "createTime": "2018-06-19T05:41:50.0041314Z",
    "version": "2018-06-01",
    "repoConfiguration": {
      "type": "FactoryVSTSConfiguration",
      "projectName": "project",
      "tenantId": "",
      "accountName": "ADF",
      "repositoryName": "repo",
      "collaborationBranch": "master",
      "rootFolder": "/",
      "lastCommitId": ""
    }
  },
  "eTag": "\"00004004-0000-0000-0000-5b28979e0000\"",
  "location": "East US",
  "tags": {
    "exampleTag": "exampleValue"
  }
}

**Review by soumyo13 — APPROVED** [review]
Lgtm

**Review by IanMadd — APPROVED** [review]

**Pull request merged**